File: generated_code/python/solution_2374_2.py
from typing import List

class Solution:
    def totalSteps(self, nums: List[int]) -> int:
        # ans tracks the maximum number of steps required overall
        ans = 0
        
        # Monotonic stack stores pairs of (value, steps_needed_to_be_eaten)
        # It is kept strictly decreasing based on the element value.
        stack = []
        
        # Iterate from right to left to determine how long each element 
        # survives relative to the larger elements to its left.
        for i in range(len(nums) - 1, -1, -1):
            cur_steps = 0
            
            # While the stack is not empty and the current element is strictly 
            # greater than the element at the top of the stack, it means the 
            # current element will eventually "eat" the stack's top element.
            while stack and nums[i] > stack[-1][0]:
                # The time required to eat the top element depends on how long 
                # that top element itself survived to eat elements to its right.
                # We need at least 1 step + the steps the top element took, 
                # or the cumulative steps calculated so far (whichever is larger).
                cur_steps = max(cur_steps + 1, stack[-1][1])
                stack.pop()
                
            # Update the global maximum steps encountered
            ans = max(ans, cur_steps)
            
            # Push the current element along with its calculated steps onto the stack
            stack.append((nums[i], cur_steps))
            
        return ans

File: generated_code/cpp/solution_2374_2.cpp
#include <vector>
#include <algorithm>
#include <stack>

using namespace std;

class Solution {
public:
    int totalSteps(vector<int>& nums) {
        int n = nums.size();
        int maxSteps = 0;
        
        // The stack will store pairs of {value, steps_needed_to_be_eaten}
        // It acts as a monotonic decreasing stack based on the element values.
        stack<pair<int, int>> st;
        
        for (int i = n - 1; i >= 0; --i) {
            int steps = 0;
            
            // While the current element is strictly greater than the element 
            // at the top of the stack, it means the current element will 
            // eventually "eat" that element.
            while (!st.empty() && nums[i] > st.top().first) {
                // The current element needs to wait for the top element to finish 
                // eating its own dependents, or it takes 1 step if it eats it directly.
                steps = max(steps + 1, st.top().second);
                st.pop();
            }
            
            // Record the maximum steps seen across the entire array
            maxSteps = max(maxSteps, steps);
            
            // Push the current element and the total steps it takes to clear 
            // its right-hand dependents.
            st.push({nums[i], steps});
        }
        
        return maxSteps;
    }
};